examples/GstHelloWorld.hs

```diff
@@ -0,0 +1,128 @@
+  putStrLn "Dynamic pad created, linking demuxer/decoder"
+  Just sinkPad <- Gst.elementGetStaticPad decoder "sink"
+  void $ Gst.padLink pad sinkPad
+  Gst.objectUnref sinkPad
+
+main :: IO ()
+main = do
+  progName <- getProgName
+  args <- getArgs
+
+  -- Initialization
+  void $ Gst.init Nothing
+
+  loop <- GLib.mainLoopNew Nothing False
+
+  -- Check input arguments
+  when (length args /= 1) $ do
+    T.hPutStrLn stderr $ "Usage: " <> pack progName <> " <Ogg/Vorbis filename>"
+    exitFailure
+  let filename = head args
+
+  -- Crate gstreamer elements
+  pipeline <- Gst.pipelineNew (Just "audio-player")
+  maySource  <- Gst.elementFactoryMake "filesrc"       (Just "file-source")
+  mayDemuxer <- Gst.elementFactoryMake "oggdemux"      (Just "ogg-demuxer")
+  mayDecoder <- Gst.elementFactoryMake "vorbisdec"     (Just "vorbis-decoder")
+  mayConv    <- Gst.elementFactoryMake "audioconvert"  (Just "converter")
+  maySink    <- Gst.elementFactoryMake "autoaudiosink" (Just "audio-output")
+
+  -- [I think this is ugly...]
+  [source,demuxer,decoder,conv,sink] <- maybe
+     (do T.hPutStrLn stderr "One element could not be created. Exiting." 
+         exitFailure 
+         undefined) 
+     return
+     (sequence [maySource, mayDemuxer, mayDecoder, mayConv, maySink])
+
+  -- Set up the pipeline
+
+  -- we set the input filename to the source element
+  setObjectPropertyString source "location" (Just $ pack filename)
+
+  -- we add a message handler
+  bus <- Gst.pipelineGetBus pipeline
+  busWatchId <- Gst.busAddWatch bus GLib.PRIORITY_DEFAULT (busCall loop)
+  Gst.objectUnref bus
+
+  -- we add all elements into the pipeline
+  -- file-source | ogg-demuxer | vorbis-decoder | converter | alsa-output
+  mapM_ (Gst.binAdd pipeline) [source, demuxer, decoder, conv, sink]
+
+  -- we link the elements together
+  -- file-source -> ogg-demuxer ~> vorbis-decoder -> converter -> alsa-output
+  void $ Gst.elementLink source demuxer
+  void $ Gst.elementLink decoder conv
+  void $ Gst.elementLink conv sink
+
+  Gst.onElementPadAdded demuxer (onPadAdded decoder)
+  
+  {- note that the demuxer will be linked to the decoder dynamically.
+     The reason is that Ogg may contain various streams (for example
+     audio and video). The source pad(s) will be created at run time,
+     by the demuxer when it detects the amount and nature of streams.
+     Therefore we connect a callback function which will be executed
+     when the "pad-added" is emitted.-}
+
+  -- Set the pipeline to "playing" state
+  putStrLn $ "Now playing: " ++ filename
+  Gst.elementSetState pipeline Gst.StatePlaying
+
+  -- Iterate
+  putStrLn "Running..."
+  #run loop
+
+  -- Out of the main loop, clean up nicely
+  putStrLn "Returned, stopping playback"
+  void $ Gst.elementSetState pipeline Gst.StateNull
+
+  -- [Is this really needed in haskell?]
+  putStrLn "Deleting pipeline"
+  Gst.objectUnref pipeline
+  GLib.sourceRemove busWatchId
+  GLib.mainLoopUnref loop
```

Comment:
There is no need to unref manually `pipeline` or `loop`, since they are owned by the Haskell bindings, and will be unreferenced when  the garbage collector removes the corresponding Haskell element.

`GLib.sourceRemove` is OK to keep. `busWatchId` is simply an integer identifying the source, and is not under the control of the GC.